### PR TITLE
Parallelized preparation and finalization in the MR mode

### PR DESCRIPTION
**Pull request by liyinan926**
1. Parallelized the serialization of `WorkUnit`s in the preparation phase.
2. Parallelized the deserialization of `TaskState`s in the finalization phase.

Signed-off-by: Yinan Li <liyinan926@gmail.com>

### CODE REVIEW

**Comment by liyinan926** [comment]
@chavdar @zliu41 can you review this?

**Comment by liyinan926** [comment]
Updated.

**Comment by zliu41** [comment]
LGTM

File: gobblin-runtime/src/main/java/gobblin/runtime/util/ParallelStateSerDeRunner.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    }));
+  }
+
+  /**
+   * Wait until the {@link ParallelStateSerDeRunner} is done running all submitted tasks.
+   *
+   * @throws InterruptedException if the calling thread is interrupted while waiting
+   * @throws ExecutionException if any of the task throws an exception during execution
+   */
+  public void awaitDone() throws InterruptedException, ExecutionException {
```
Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/util/ParallelStateSerDeRunner.java]: Can't we have the awaitDone() logic as part of close() ? This way we don't have to worry if awaitDone() has been called before close().

**Comment by liyinan926** [comment]
Updated.

**Comment by chavdar** [comment]
LGTM.

**Pull request merged**